crates/sui-indexer/src/apis/governance_api_v2.rs

```diff
@@ -5,27 +5,153 @@
 #![allow(unused_variables)]
 #![allow(dead_code)]
 
-use crate::store::PgIndexerStoreV2;
+use crate::errors::{Context, IndexerError};
+use crate::models_v2::epoch::StoredEpochInfo;
+use crate::store::IndexerStoreV2;
+use crate::PgConnectionPool;
 use async_trait::async_trait;
 use jsonrpsee::core::RpcResult;
 use jsonrpsee::RpcModule;
 
 use sui_json_rpc::api::GovernanceReadApiServer;
 use sui_json_rpc::SuiRpcModule;
-use sui_json_rpc_types::SuiCommittee;
 use sui_json_rpc_types::{DelegatedStake, ValidatorApys};
+use sui_json_rpc_types::{EpochInfo, SuiCommittee};
 use sui_open_rpc::Module;
-use sui_types::base_types::{ObjectID, SuiAddress};
+use sui_types::base_types::{ObjectID, SequenceNumber, SuiAddress, VersionNumber};
+use sui_types::committee::EpochId;
+use sui_types::object::Object;
+use sui_types::storage::ObjectStore;
 use sui_types::sui_serde::BigInt;
 use sui_types::sui_system_state::sui_system_state_summary::SuiSystemStateSummary;
+use sui_types::sui_system_state::SuiSystemStateTrait;
 
+use crate::schema_v2::{epochs, objects};
+use crate::store::diesel_marco::read_only_blocking;
+
+use diesel::dsl::{count, max};
+use diesel::pg::PgConnection;
+// use diesel::sql_types::{BigInt, VarChar};
+use diesel::upsert::excluded;
+use diesel::ExpressionMethods;
+use diesel::{OptionalExtension, QueryableByName};
+use diesel::{QueryDsl, RunQueryDsl};
+
+#[derive(Clone)]
 pub(crate) struct GovernanceReadApiV2 {
-    pg_store: PgIndexerStoreV2,
+    blocking_cp: PgConnectionPool,
 }
 
 impl GovernanceReadApiV2 {
-    pub fn new(pg_store: PgIndexerStoreV2) -> Self {
-        Self { pg_store }
+    pub fn new(blocking_cp: PgConnectionPool) -> Self {
+        Self { blocking_cp }
+    }
+
+    fn get_reference_gas_price(&self) -> Result<u64, IndexerError> {
+        let epoch_info = self.get_epoch_info(None)?;
+        Ok(epoch_info.reference_gas_price.unwrap())
+    }
+
+    fn get_latest_sui_system_state(&self) -> Result<SuiSystemStateSummary, IndexerError> {
+        let system_state: SuiSystemStateSummary =
+            sui_types::sui_system_state::get_sui_system_state(self)?
+                .into_sui_system_state_summary();
+        Ok(system_state)
+    }
+
+    fn get_epoch_info(&self, epoch: Option<EpochId>) -> Result<EpochInfo, IndexerError> {
+        let epoch: StoredEpochInfo = read_only_blocking!(&self.blocking_cp, |conn| {
+            if let Some(epoch) = epoch {
+                epochs::dsl::epochs
+                    .filter(epochs::epoch.eq(epoch as i64))
+                    .limit(1)
+                    .first(conn)
```

Comment:
question: could this also be affixed with .optional() for scenarios where an epoch that doesn't exist yet is referenced?